### PR TITLE
posix: redo socket API to use conn

### PR DESCRIPTION
**Pull request by authmillenon**
Depends on ~~#3611~~ (merged), ~~#3612~~ (merged), ~~#3615~~ (merged) and ~~#3660~~ (merged).

### CODE REVIEW

**Comment by BytesGalore** [comment]
oh that's not really what I want, sry.

File: sys/posix/sockets/posix_sockets.c [hunk 1]
```diff
@@ -0,0 +1,624 @@
...
+        case AF_INET6:
+#endif
+#if CONN_IPV4 || CONN_IPV6
+            s->domain = domain;
+            s->type = type;
+            if ((s->protocol = _choose_ipproto(type, protocol)) < 0) {
+                res = -1;
+            }
+            break;
+#endif
```
Comment (authmillenon) [inline sys/posix/sockets/posix_sockets.c]: From https://github.com/authmillenon/RIOT/commit/902d305123e6abb27ae4924ac8d5f37c75e5e6e2#commitcomment-12756195:

> how about:
>
>     case AF_INET:
>     case AF_INET6:
>         s->domain = domain;
>         s->type = type;
>         if ((s->protocol = _choose_ipproto(type, protocol)) < 0) {
>             res = -1;
>         }
>         break;

That's different from the current code in that this code segment will be entered even if the associated connection modules are not there. But I can put the `EAFNOSUPPORT` check for them to the top.

Comment (BytesGalore) [inline sys/posix/sockets/posix_sockets.c]: ahh `domain` may be `AF_INET` or `AF_INET6` even if `CONN_IPV4 || CONN_IPV6` are not available, the forget my noise :)

**Comment by authmillenon** [comment]
Rebased to current master

**Comment by authmillenon** [comment]
Rebased to current master.

**Comment by authmillenon** [comment]
> oh that's not really what I want, sry.

@BytesGalore what did you mean with that?

**Comment by authmillenon** [comment]
Rebased to current dependencies

**Comment by BytesGalore** [comment]
> @BytesGalore what did you mean with that?

nvm. @OlegHahm pointed me that I commented on commits from this PR and not on the PR, so that was something I didn't wanted :)

**Comment by authmillenon** [comment]
Rebased to current master and dependencies.

**Comment by authmillenon** [comment]
And now for real.

**Comment by authmillenon** [comment]
The example application is "done". Receiving does not work properly yet, but I'm investigating.

**Comment by emmanuelsearch** [comment]
great. Let me know when I should test.

**Comment by authmillenon** [comment]
Rebased to current master and dependencies and fixed the example application. Should work now. @emmanuelsearch: ready for testing

**Comment by authmillenon** [comment]
Rebased to current master and dependencies.

**Comment by authmillenon** [comment]
Rebased to current #conn

**Comment by authmillenon** [comment]
Rebased to current master. No longer waiting on other PR.

File: examples/posix_sockets/udp.c [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        socklen_t src_len = sizeof(struct sockaddr_in6);
+        if ((res = recvfrom(server_socket, server_buffer, sizeof(server_buffer), 0,
+                            (struct sockaddr *)&src, &src_len)) < 0) {
+            perror("Error on receive");
+        }
+        else if (res == 0) {
+            puts("Peer did shut down");
+        }
+        else {
+            puts(server_buffer);
```
Comment (OlegHahm) [inline examples/posix_sockets/udp.c]: A prefix like `Packet received...` would be helpful, I guess.

Comment (authmillenon) [inline examples/posix_sockets/udp.c]: Addressed

File: sys/posix/include/netinet/in.h [hunk 3]
```diff
@@ -25,135 +25,244 @@
...
 
-    /**
-     * IPv6 Address represented as sequence of 8-bit numbers. Member of
-     * struct in6_addr.
-     */
-#define s6_addr     __in6_u.uint8
+/**
+ * IPv6 wildcard address.
+ */
+#define IN6ADDR_ANY_INIT        {{0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0}}
```
Comment (OlegHahm) [inline sys/posix/include/netinet/in.h]: Why not simply `#define IN6ADDR_ANY_INIT IPV6_ADDR_UNSPECIFIED`?

File: sys/posix/include/netinet/in.h [hunk 4]
```diff
@@ -25,135 +25,244 @@
...
 
-    /**
-     * IPv6 Address represented as sequence of 16-bit numbers. Member of
-     * struct in6_addr.
-     */
-#define s6_addr16   __in6_u.uint16
+/**
+ * IPv6 loopback address.
+ */
+#define IN6ADDR_LOOPBACK_INIT   {{0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1}}
```
Comment (OlegHahm) [inline sys/posix/include/netinet/in.h]: and `IPV6_ADDR_LOOPBACK` here?

**Comment by emmanuelsearch** [comment]
When I run the example and I try to start the udp server with `udp server start 40` I get an error `Error: server already running` whatever port number I try (even higher port numbers e.g. above 1024)

File: sys/posix/include/netinet/in.h (hunk 3 above)
Comment (authmillenon) [inline sys/posix/include/netinet/in.h]: IIRC they may have different formats, regarding numbers of braces around them.

**Comment by emmanuelsearch** [comment]
When I try to compile the example on OS X, I get the following errors:

```
In file included from /Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/sys/posix/include/arpa/inet.h:34:
/Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/sys/posix/include/sys/bytes.h:40:9: error: 
      'htonl' macro redefined [-Werror,-Wmacro-redefined]
#define htonl(hostlong)     HTONL(hostlong)
        ^
/usr/include/sys/_endian.h:136:9: note: previous definition is here
#define htonl(x)        __DARWIN_OSSwapInt32(x)
        ^
In file included from /Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/examples/posix_sockets/udp.c:26:
In file included from /Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/sys/posix/include/arpa/inet.h:34:
/Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/sys/posix/include/sys/bytes.h:53:9: error: 
      'htons' macro redefined [-Werror,-Wmacro-redefined]
#define htons(hostshort)    HTONS(hostshort)
        ^
/usr/include/sys/_endian.h:133:9: note: previous definition is here
#define htons(x)        __DARWIN_OSSwapInt16(x)
        ^
In file included from /Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/examples/posix_sockets/udp.c:26:
In file included from /Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/sys/posix/include/arpa/inet.h:34:
/Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/sys/posix/include/sys/bytes.h:66:9: error: 
      'ntohl' macro redefined [-Werror,-Wmacro-redefined]
#define ntohl(netlong)      NTOHL(netlong)
        ^
/usr/include/sys/_endian.h:135:9: note: previous definition is here
#define ntohl(x)        __DARWIN_OSSwapInt32(x)
        ^
In file included from /Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/examples/posix_sockets/udp.c:26:
In file included from /Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/sys/posix/include/arpa/inet.h:34:
/Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/sys/posix/include/sys/bytes.h:79:9: error: 
      'ntohs' macro redefined [-Werror,-Wmacro-redefined]
#define ntohs(netshort)     NTOHS(netshort)
        ^
/usr/include/sys/_endian.h:132:9: note: previous definition is here
#define ntohs(x)        __DARWIN_OSSwapInt16(x)
        ^
4 errors generated
```

File: sys/posix/include/netinet/in.h (hunk 3 above)
Comment (OlegHahm) [inline sys/posix/include/netinet/in.h]:
```c
#define IPV6_ADDR_UNSPECIFIED               {{ 0x00, 0x00, 0x00, 0x00, \
                                               0x00, 0x00, 0x00, 0x00, \
                                               0x00, 0x00, 0x00, 0x00, \
                                               0x00, 0x00, 0x00, 0x00 }}

**Comment by authmillenon** [comment]
> When I run the example and I try to start the udp server with udp server start 40 I get an error Error: server already running whatever port number I try (even higher port numbers e.g. above 1024)

I don't get that error on native. Which platform are you using?

**Comment by emmanuelsearch** [comment]
I use native.

**Comment by authmillenon** [comment]
Try the fix I provided regarding the compile error.

**Comment by emmanuelsearch** [comment]
will try the fix for the compile error.

**Comment by emmanuelsearch** [comment]
for the running error: it seems that starting the server only works on tap0 node (not on tap 1).

**Comment by authmillenon** [comment]
works on tap1 for me.

**Comment by authmillenon** [comment]
@emmanuelsearch but it worked when you first tested it?

**Comment by OlegHahm** [comment]
Works on tap1 here, too, but I can confirm that it fails on @emmanuelsearch's VM. ;-)

**Comment by emmanuelsearch** [comment]
Compiles on OS X with the latest commit.

**Comment by authmillenon** [comment]
That's weird… the tap interface should have nothing to do with the `server_socket` variable (this one is checked on `udp_start_server()`) in the binary. Can you maybe output it on `udp_start_server()`? It should be `-1` before the first call and then stay the same (the fd of the socket) afterwards

**Comment by emmanuelsearch** [comment]
Starting the UDP server did not work the first time I tried on the tap1 node, but:
- it worked on the tap0 node and I could send from tap1 to tap 0
- I then recompiled and restarted and it worked both ways;

File: sys/posix/include/netinet/in.h (hunk 3 above)
Comment (authmillenon) [inline sys/posix/include/netinet/in.h]: Tried it, worked => addressed

File: sys/posix/include/netinet/in.h (hunk 4 above)
Comment (authmillenon) [inline sys/posix/include/netinet/in.h]: Addressed

**Comment by emmanuelsearch** [comment]
Maybe the READ ME could be more complete -- especially for programmers expected above a POSIX socket in this context. E.g. in EXAMPLE OUTPUT, indicate that `ifconfig` gives you the IP address, and say that the UDP port number used here is not to be confused with the port used for the shell indicated in USAGE. In USAGE, indicate that on native, no need to indicate a port.

**Comment by authmillenon** [comment]
> In USAGE, indicate that on native, no need to indicate a port.

Since when? If you want to use a different tap interface apart from `tap0` you need to inticate a port.

**Comment by emmanuelsearch** [comment]
> In USAGE, indicate that on native, no need to indicate a port.

oops my bad. Ignore ;)

**Comment by authmillenon** [comment]
Update README.

**Comment by emmanuelsearch** [comment]
Looks good. ACK.

**Comment by authmillenon** [comment]
Squashed

**Comment by authmillenon** [comment]
Fixed issues with `coap` app and added low-memory boards to `BOARD_INSUFFICIENT_MEMORY`.

**Comment by authmillenon** [comment]
Another (bug) one bites the dust. This time it was Arduino with its pesky 16-bit `int`s again

**Comment by authmillenon** [comment]
(hopefully) last one. `tests/coap` is due to `gnrc` too big for `stm32f0discovery`.

**Comment by authmillenon** [comment]
Let's give it one last round for today, shall we (blacklisted `msp430` and `avr` boards for `tests/coap` again, since they have problems with libcoaps `:x` notation in structs and added include to `kernel_types.h` to `sys/posix.h` to have `ssize_t` on `msp430`)

**Comment by OlegHahm** [comment]
The last comment sounded like a question...

**Comment by authmillenon** [comment]
It was a rhetorical question intended to mask my frustration about the last minutes of this PR.

**Pull request merged**